README.markdown
WhereMaJamzAt
==========
<i>Paul Pietkiewicz <paul.pietkiewicz at acm dot org> </i>


Intro:
-------

I needed to try out some web scrapping. You see there is this radio I like to listen to. Or at least used to like
to listen to it... when it was available outside the US, and I thought it might be nice to create a playlist
of songs so I could "emulate" the experience. Maybe without the useless talk. So this is how WhereMaJamzAt.

BTW: I condone piracy, so this is just for educational purposes.


Where:
-------
[https://github.com/platten/WhereMaJamzAt](https://github.com/platten/WhereMaJamzAt)


Prereqs:
--------
- Python 2.7
- [ID3](http://id3-py.sourceforge.net/)
- [Beautiful Soup 4](http://www.crummy.com/software/BeautifulSoup/)
- [mplayer](http://www.mplayerhq.hu/design7/dload.html)


Licensing:
==========
See da LICENSE

CHANGES.markdown
CHANGES:
========


0.888
-----
 - Initial version


src/wheremajamzat/utils.py
"""utils.py: Supplementary utility functions for WhereMyJamz@"""

from urllib import urlopen
import datetime
import re
import subprocess
import os
from bs4 import BeautifulSoup

stations = ['kroq', 'whfs', 'v103', 'whfs']
template = 'http://{station}.radio.com/playlist/{year}/{month}/{day}/'


def identify_songs(datetime_obj=datetime.datetime.now(), station=stations[0]):
    songs = []

    template_dict = {'station': station,
                     'year': str(datetime_obj.year),
                     'month': "%02d" % datetime_obj.month,
                     'day': "%02d" % datetime_obj.day}
    f = urlopen(template.format(**template_dict))
    soup = BeautifulSoup(f)
    tracks = soup.findAll('div', attrs='track_info')
    for track in tracks[9:]:
        title = track.findChild('div', attrs='track_title').text
        artist = track.findChild('div', attrs='track_artist').text
        songs.append({'title': title, 'artist': artist})
    return songs


def get_song_length(track_path):
    song_length = subprocess.check_output('midentify.sh "%s" '
                                '| grep ID_LENGTH | cut -d"=" -f2' %
                                track_path, shell=True)
    return int(eval(song_length.rstrip()))


def write_playlist(playlist_path, songs):
    fp = open(playlist_path, 'w')
    fp.write('#EXTM3U\n')
    for song in songs:
        fp.write('#EXTINF:%d,%s - %s\n' % (get_song_length(song['path']),
                                           song['name'], song['artist']))
        fp.write(song['path'] + '\n')
    fp.close()
    return playlist_path


def get_track_name(title, artist):
    return re.sub(' ', '_', re.sub('"', '\"', "%s-%s.mp3" % (artist, title)))


def get_directory_name(station, date):
    directory = "%s-%s" % (station, date.strftime('%m-%d-%y'))
    dir_path = os.path.abspath(os.path.join(os.path.curdir, directory))
    if not os.path.exists(dir_path):
        os.mkdir(dir_path)
    return dir_path


src/wheremajamzat/grooveshark.py
"""grooveshark_downloader.py: mp3 file downloader based on the wonderful
                            work of George Stephanos's groove-dl
                            http://github.com/jacktheripper51/groove-dl"""


import httplib
import StringIO
import hashlib
import uuid
import random
import string
import sys
import gzip
import datetime
import os
import subprocess
import time
from ID3 import *
from utils import get_song_length, get_track_name

if sys.version_info[1] >= 6:
    import json
else:
    import simplejson as json

_useragent = "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/535.11 " \
             "(KHTML, like Gecko) Chrome/17.0.963.56 Safari/535.11"
_referer = "http://grooveshark.com/JSQueue.swf?20120220.01"
wget_template = 'wget --user-agent="%s" --referer=%s ' \
                '--post-data=streamKey=%s -O "%s" "http://%s/stream.php"'
_token = None

h = {}
h["country"] = {}
h["country"]["CC1"] = "0"
h["country"]["CC2"] = "0"
h["country"]["CC3"] = "0"
h["country"]["CC4"] = "0"
h["country"]["ID"] = "1"
h["privacy"] = 0
h["session"] = None
h["uuid"] = str.upper(str(uuid.uuid4()))


def prepToken(method, secret):
    rnd = (''.join(random.choice(string.hexdigits) for x in range(6))).lower()
    return rnd + hashlib.sha1(method + ":" + _token + secret + rnd).hexdigest()


def getToken():
    global h, _token
    p = {}
    p["parameters"] = {}
    p["parameters"]["secretKey"] = hashlib.md5(h["session"]).hexdigest()
    p["method"] = "getCommunicationToken"
    p["header"] = h
    p["header"]["client"] = "htmlshark"
    p["header"]["clientRevision"] = "20120220"
    conn = httplib.HTTPSConnection("grooveshark.com")
    conn.request("POST", "/more.php", json.JSONEncoder().encode(p),
                                    {"User-Agent": _useragent,
                                     "Referer": _referer,
                                     "Content-Type": "",
                                     "Accept-Encoding": "gzip",
                                     "Cookie": "PHPSESSID=" + h["session"]})
    _token = json.JSONDecoder().decode(gzip.GzipFile(
                    fileobj=(StringIO.StringIO(
                        conn.getresponse().read()))).read())["result"]


def getResultsFromSearch(query, what="Songs"):
    p = {}
    p["parameters"] = {}
    p["parameters"]["type"] = what
    p["parameters"]["query"] = query
    p["header"] = h
    p["header"]["client"] = "htmlshark"
    p["header"]["clientRevision"] = "20120220"
    p["header"]["token"] = prepToken("getResultsFromSearch", ":jayLikeWater:")
    p["method"] = "getResultsFromSearch"
    conn = httplib.HTTPConnection("grooveshark.com")
    conn.request("POST", "/more.php?" + p["method"],
            json.JSONEncoder().encode(p),
            {"User-Agent": _useragent,
             "Referer": "http://grooveshark.com/",
             "Content-Type": "application/json",
             "Accept-Encoding": "gzip",
             "Cookie": "PHPSESSID=" + h["session"]})
    j = json.JSONDecoder().decode(gzip.GzipFile(
                    fileobj=(StringIO.StringIO(
                        conn.getresponse().read()))).read())
    try:
        return j["result"]["result"]["Songs"]
    except:
        return j["result"]["result"]


def artistGetSongsEx(id, isVerified):
    p = {}
    p["parameters"] = {}
    p["parameters"]["artistID"] = id
    p["parameters"]["isVerifiedOrPopular"] = isVerified
    p["header"] = h
    p["header"]["client"] = "htmlshark"
    p["header"]["clientRevision"] = "20120220"
    p["header"]["token"] = prepToken("artistGetSongsEx", ":jayLikeWater:")
    p["method"] = "artistGetSongsEx"
    conn = httplib.HTTPConnection("grooveshark.com")
    conn.request("POST", "/more.php?" + p["method"],
                    json.JSONEncoder().encode(p),
                    {"User-Agent": _useragent,
                     "Referer": _referer,
                     "Content-Type": "",
                     "Accept-Encoding": "gzip",
                     "Cookie": "PHPSESSID=" + h["session"]})
    return json.JSONDecoder().decode(gzip.GzipFile(
                        fileobj=(StringIO.StringIO(
                            conn.getresponse().read()))).read())


def getStreamKeyFromSongIDEx(id):
    p = {}
    p["parameters"] = {}
    p["parameters"]["mobile"] = "false"
    p["parameters"]["prefetch"] = "false"
    p["parameters"]["songIDs"] = id
    p["parameters"]["country"] = h["country"]
    p["header"] = h
    p["header"]["client"] = "jsqueue"
    p["header"]["clientRevision"] = "20120220.01"
    p["header"]["token"] = prepToken("getStreamKeysFromSongIDs",
                                            ":bangersAndMash:")
    p["method"] = "getStreamKeysFromSongIDs"
    conn = httplib.HTTPConnection("grooveshark.com")
    conn.request("POST", "/more.php?" + p["method"],
                            json.JSONEncoder().encode(p),
                        {"User-Agent": _useragent,
                         "Referer": _referer,
                         "Accept-Encoding": "gzip",
                         "Content-Type": "",
                         "Cookie": "PHPSESSID=" + h["session"]})
    j = json.JSONDecoder().decode(gzip.GzipFile(
                            fileobj=(StringIO.StringIO(
                                conn.getresponse().read()))).read())
    return j


def header_cb(buf):
    global h
    if "PHPSESSID" in buf:
        buf = buf.split(' ')
        h["session"] = buf[1][10:-1]


def init():
    conn = httplib.HTTPConnection("grooveshark.com")
    conn.request("HEAD", "", headers={"User-Agent": _useragent})
    res = conn.getresponse()
    cookie = res.getheader("set-cookie").split(";")
    h["session"] = cookie[0][10:]


def download_songs(download_directory, song_list, aggressive=False):
    songs = []
    init()
    getToken()
    if not aggressive:
        time.sleep(3)
    try:
        for song in song_list:
            print "Searching for: %s by %s" % (song['title'], song['artist'])
            s = getResultsFromSearch("%s %s" % (song['title'], song['artist']))
            stream = getStreamKeyFromSongIDEx(s[0]["SongID"])
            for k, v in stream["result"].iteritems():
                stream = v
            title = s[0]["SongName"]
            artist = s[0]["ArtistName"]
            track_path = os.path.join(download_directory,
                                    get_track_name(title, artist))
            start_time = datetime.datetime.now()
            if not os.path.exists(track_path):
                wait_flag = True
                wgets = wget_template % (_useragent, _referer,
                                stream["streamKey"], track_path, stream["ip"])
                p = subprocess.Popen(wgets, shell=True)
                p.wait()
                try:
                    id3info = ID3(track_path)
                    id3info['TITLE'] = title
                    id3info['ARTIST'] = artist
                    id3info.write()
                    print 'ID3 tags written'
                except InvalidTagError, message:
                    print "Invalid ID3 tag:", message
            else:
                print 'File %s already exists' % track_path
            wait_flag = False
            if not aggressive and song != song_list[-1]:
                if wait_flag:
                    track_length = datetime.timedelta(
                                seconds=get_song_length(track_path))
                    end_time = datetime.datetime.now()
                    to_wait = track_length - (end_time - start_time)
                    to_wait = to_wait.total_seconds() + 2
                if not wait_flag:
                    to_wait = random.randrange(2, 6)
                print "Sleeping for %d seconds" % to_wait
                time.sleep(to_wait)
            songs.append({'path': track_path, 'name': title, 'artist': artist})

    except TypeError:
        print "Just got banned by grooveshark, stopping download "

    return songs


src/wheremyjamzat/wheremyjamzat.py
#!/usr/bin/env python

"""wheremyjamzat.py: Get those jamz"""


__author__    = "Paul Pietkiewicz"
__copyright__ = "Copyright 2012, Paul Pietkiewicz"
__email__     = "pawel.pietkiewicz@acm.org"
__version__   = '0.888'

import sys
import argparse
import os
import datetime

from utils import get_directory_name, write_playlist, stations, identify_songs
from grooveshark import download_songs

def main():
    """CLI Driver"""
    parser = argparse.ArgumentParser(description='Music downloader and playlist generator', \
                                     prog='wheremyjamzat.py')
    parser.add_argument('station',
        help='Radio station name', choices=stations)
    now = datetime.datetime.now()

    parser.add_argument('-y', help='Year', action="store", dest="year", type=int, choices=xrange(now.year -2, now.year), default=now.year)
    parser.add_argument('-m', help='Month', action="store", dest="month", type=int, choices=xrange(1, 12), default=now.month)
    parser.add_argument('-d', help='Day', action="store", dest="day", type=int, choices=xrange(1, 31), default=now.day)
    parser.add_argument('-n', help='Number of songs to download', action="store", dest="num_songs", type=int, default=-1)

    cli_args = parser.parse_args()
    provided_date = datetime.datetime(year=cli_args.year, month=cli_args.month, day=cli_args.day)
    if provided_date > now:
        raise argparse.ArgumentError('Provided date cannot be in the future')
	if cli_args.num_songs < -1:
		raise argparse.ArgumentError('Invalid number of songs')

    song_list = identify_songs(provided_date, cli_args.station)
    destination_path = os.path.join(os.path.curdir,
                    get_directory_name(cli_args.station, provided_date))

    if cli_args.num_songs != -1 and cli_args.num_songs < len(song_list):
	    song_list = song_list[:cli_args.num_songs]

    downloaded_songs = download_songs(destination_path, song_list)
    playlist_path = write_playlist(os.path.join(destination_path, 'playlist.m3u'),
        downloaded_songs)
    print "Playlist created: %s" % playlist_path
    print "Done!"

if __name__ == "__main__":
    main()
    sys.exit(0)


src/wheremajamzat/wheremajamzat.py
#!/usr/bin/env python

"""wheremajamzat.py: Get those jamz"""


__author__    = "Paul Pietkiewicz"
__copyright__ = "Copyright 2012, Paul Pietkiewicz"
__email__     = "pawel.pietkiewicz@acm.org"
__version__   = '0.888'

import sys
import argparse
import os
import datetime

from utils import get_directory_name, write_playlist, stations, identify_songs
from grooveshark import download_songs


def main():
    """CLI Driver"""
    parser = argparse.ArgumentParser(description='Music downloader and ' \
                                                 'playlist generator', \
                                     prog='wheremajamzat.py')
    parser.add_argument('station',
        help='Radio station name', choices=stations)
    now = datetime.datetime.now()

    parser.add_argument('-y', help='Year', action="store", dest="year",
        type=int, choices=xrange(now.year - 2, now.year + 1), default=now.year)
    parser.add_argument('-m', help='Month', action="store",
        dest="month", type=int, choices=xrange(1, 12), default=now.month)
    parser.add_argument('-d', help='Day', action="store",
        dest="day", type=int, choices=xrange(1, 31), default=now.day)
    parser.add_argument('-n', help='Number of songs to download',
        action="store", dest="num_songs", type=int, default=-1)
    parser.add_argument('-A',
        help='Agressive downloading (warning, can get you banned)',
        action="store_true", dest="aggressive")

    cli_args = parser.parse_args()
    provided_date = datetime.datetime(year=cli_args.year, month=cli_args.month,
                                    day=cli_args.day)
    if provided_date > now:
        raise argparse.ArgumentError('Provided date cannot be in the future')
    if cli_args.num_songs < -1:
        raise argparse.ArgumentError('Invalid number of songs')

    song_list = identify_songs(provided_date, cli_args.station)
    destination_path = os.path.join(os.path.curdir,
                    get_directory_name(cli_args.station, provided_date))

    if cli_args.num_songs != -1 and cli_args.num_songs < len(song_list):
        song_list = song_list[:cli_args.num_songs]

    downloaded_songs = download_songs(destination_path, song_list,
                                    aggressive=cli_args.aggressive)
    playlist_path = write_playlist(os.path.join(destination_path,
                                                    'playlist.m3u'),
        downloaded_songs)
    print "Playlist created: %s" % playlist_path
    print "Done!"

if __name__ == "__main__":
    main()
    sys.exit(0)
